### PR TITLE
Hyperparameter Optimization for Trainable Surrogates

### PR DESCRIPTION
**Pull request by jduerholt**
This PR adds the possibility to optimize hyperparameters of bofire surrogates in an easy manner with bofire. For this purpose, trainable surrogates can have a configuration object attached in which the hyperparameters and its ranges etc. are defined, together with a method how to update them. 

Changing this configuration object which is part of the surrogate's data model allows to tweak the applied hyperparameter optimization. 

Furthermore, this PR adds:
- `FactorialStrategy` for purely discrete/categorical domains
- LinearSurrogate to easily setup linear models

In a next step, one could enable automatic hyperparameter optimization also in strategies.

### CODE REVIEW

File: bofire/data_models/surrogates/single_task_gp.py [hunk 1]
```diff
@@ -1,19 +1,98 @@
...
+    hyperstrategy: Literal[
+        "FactorialStrategy", "SoboStrategy", "RandomStrategy"
+    ] = "FactorialStrategy"
+
+    @staticmethod
+    def _update_hyperparameters(
+        surrogate_data: "SingleTaskGPSurrogate", hyperparameters: pd.Series
+    ):
+        def matern_25(ard: bool, lenghtscale_prior: AnyPrior) -> MaternKernel:
+            return MaternKernel(nu=2.5, lengthscale_prior=lenghtscale_prior, ard=ard)
```
Comment (R-M-Lee) [inline bofire/data_models/surrogates/single_task_gp.py:52]: @jduerholt typo here: lenghtscale_prior --> lengthscale_prior

Comment (jduerholt) [reply bofire/data_models/surrogates/single_task_gp.py]: Thanks for spotting, I just fixed it.

**Comment by R-M-Lee** [comment]
Looks great by the way. What do you think about allowing GroupKFold in addition to KFold, or even allowing the user to specify their own cross validation iterator?

**Comment by bertiqwerty** [comment]
If Robert says it looks great, we can approve it I guess :)

**Review by bertiqwerty — APPROVED** [review]

**Pull request merged**